pkg/controller/certificates/issuing/issuing_controller_test.go

```diff
@@ -932,3 +964,12 @@ func TestIssuingController(t *testing.T) {
 		})
 	}
 }
+
+// We don't need to full bundle, just a simple CertificateRequest.
+func createCertificateRequestOrPanic(crt *cmapi.Certificate) *cmapi.CertificateRequest {
```

Comment:
I think this can be replaced with MustCreateCryptoBundle

Reply:
Good point; I struggled a bit finding a way to pass a `*testing.T` to [MustCreateCryptoBundle](https://github.com/jetstack/cert-manager/blob/7fbdd6487646e812fe74c0c05503805b5d9d4751/pkg/controller/certificates/internal/test/test.go#L74-L80); for example I could add `*testing.T` to all the test cases that use the field `builder`:

```diff
--- a/pkg/controller/certificates/issuing/issuing_controller_test.go
+++ b/pkg/controller/certificates/issuing/issuing_controller_test.go
@@ -49,7 +49,7 @@ func testLocalTemporarySignerFn(b []byte) localTemporarySignerFn {
 
 func TestIssuingController(t *testing.T) {
 	type testT struct {
-		builder *testpkg.Builder
+		builder func(*testing.T) *testpkg.Builder
 
 		certificate *cmapi.Certificate
 
@@ -176,32 +176,34 @@ func TestIssuingController(t *testing.T) {
 
 		"if certificate is in Issuing state, one CertificateRequest, but has failed and does not match the certificate spec, do nothing": {
 			certificate: exampleBundle.Certificate,
-			builder: &testpkg.Builder{
-				CertManagerObjects: []runtime.Object{
-					issuingCert.DeepCopy(),
-					gen.CertificateRequestFrom(createCertificateRequestOrPanic(gen.CertificateFrom(issuingCert,
-						gen.SetCertificateDNSNames("foo.com"),
-					)), gen.AddCertificateRequestAnnotations(map[string]string{
-						cmapi.CertificateRequestRevisionAnnotationKey: "2", // Current Certificate revision=1
-					}), gen.SetCertificateRequestStatusCondition(cmapi.CertificateRequestCondition{
-						Type:    cmapi.CertificateRequestConditionReady,
-						Status:  cmmeta.ConditionFalse,
-						Reason:  cmapi.CertificateRequestReasonFailed,
-						Message: "The certificate request failed because of reasons",
-					})),
-				},
-				KubeObjects: []runtime.Object{
-					&corev1.Secret{
-						ObjectMeta: metav1.ObjectMeta{
-							Name:      nextPrivateKeySecretName,
-							Namespace: exampleBundle.Certificate.Namespace,
-						},
-						Data: map[string][]byte{
-							corev1.TLSPrivateKeyKey: exampleBundle.PrivateKeyBytes,
+			builder: func(t *testing.T) *testpkg.Builder {
+				return &testpkg.Builder{
+					CertManagerObjects: []runtime.Object{
+						issuingCert.DeepCopy(),
+						gen.CertificateRequestFrom(createCertificateRequestOrPanic(gen.CertificateFrom(issuingCert,
+							gen.SetCertificateDNSNames("foo.com"), // Mismatch since the cert has "example.com"
+						)), gen.AddCertificateRequestAnnotations(map[string]string{
+							cmapi.CertificateRequestRevisionAnnotationKey: "2", // Current Certificate revision=1
+						}), gen.SetCertificateRequestStatusCondition(cmapi.CertificateRequestCondition{
+							Type:    cmapi.CertificateRequestConditionReady,
+							Status:  cmmeta.ConditionFalse,
+							Reason:  cmapi.CertificateRequestReasonFailed,
+							Message: "The certificate request failed because of reasons",
+						})),
+					},
+					KubeObjects: []runtime.Object{
+						&corev1.Secret{
+							ObjectMeta: metav1.ObjectMeta{
+								Name:      nextPrivateKeySecretName,
+								Namespace: exampleBundle.Certificate.Namespace,
+							},
+							Data: map[string][]byte{
+								corev1.TLSPrivateKeyKey: exampleBundle.PrivateKeyBytes,
+							},
 						},
 					},
```

This meant updating all the test cases with this new signature 😄

What do you think?

Reply:
Is that because you want to pass the `t` from the sub-test rather than the top-level `t` ? 
I don't think that's necessary, becasue at this point in the code we're **constructing** the testcases rather than executing them so it makes sense to fail with the top-level `t`.

Reply:
You read my mind! 😇 

Maarjte also mentioned this in https://github.com/jetstack/cert-manager/pull/3478#discussion_r529572372 and https://github.com/jetstack/cert-manager/pull/3476#discussion_r534953384, and you are both right: this is just the "construction" of the testing fixtures, so we don't really need to bother with grabbing the `t` at the subtest level, the top-level is fine. 😅

Thanks for the detailed comment, that was really helpful!!